src/cdk-experimental/scrolling/virtual-for-of.ts

```diff
@@ -256,82 +252,64 @@ export class CdkVirtualForOf<T> implements CollectionViewer, DoCheck, OnDestroy
-    // TODO(mmalerba): Currently we remove every view and then re-insert it in the correct place.
-    // It would be better to generate the minimal set of remove & inserts to get to the new list
-    // instead.
-
-    // Detach all of the views and add them into an array to preserve their original order.
-    const previousViews: (EmbeddedViewRef<CdkVirtualForOfContext<T>> | null)[] = [];
-    let i = this._viewContainerRef.length;
-    while (i--) {
-      previousViews.unshift(
-          this._viewContainerRef.detach()! as EmbeddedViewRef<CdkVirtualForOfContext<T>>);
-    }
-
-    // Mark the removed indices so we can recycle their views.
-    changes.forEachRemovedItem(record => {
-      this._templateCache.push(previousViews[record.previousIndex!]!);
-      previousViews[record.previousIndex!] = null;
-    });
-
-    // Queue up the newly added items to be inserted, recycling views from the cache if possible.
-    const insertTuples: RecordViewTuple<T>[] = [];
-    changes.forEachAddedItem(record => {
-      insertTuples[record.currentIndex!] = {record, view: this._templateCache.pop()};
-    });
-
-    // Queue up moved items to be re-inserted.
-    changes.forEachMovedItem(record => {
-      insertTuples[record.currentIndex!] = {record, view: previousViews[record.previousIndex!]!};
-      previousViews[record.previousIndex!] = null;
+    // Rearrange the views to put them in the right location.
+    changes.forEachOperation(
+        (record: IterableChangeRecord<T>, adjustedPreviousIndex: number, currentIndex: number) => {
+          if (record.previousIndex == null) {  // Item added.
+            const view = this._getViewForNewItem();
+            this._viewContainerRef.insert(view, currentIndex);
+            view.context.$implicit = record.item;
+          } else if (currentIndex == null) {  // Item removed.
+            this._cacheView(this._viewContainerRef.detach(adjustedPreviousIndex) as
+                EmbeddedViewRef<CdkVirtualForOfContext<T>>);
+          } else {  // Item moved.
+            const view = this._viewContainerRef.get(adjustedPreviousIndex) as
+                EmbeddedViewRef<CdkVirtualForOfContext<T>>;
+            this._viewContainerRef.move(view, currentIndex);
+            view.context.$implicit = record.item;
+          }
+        });
+
+    // Update $implicit for any items that had an identity change.
+    changes.forEachIdentityChange((record: any) => {
+      const view = this._viewContainerRef.get(record.currentIndex) as
+          EmbeddedViewRef<CdkVirtualForOfContext<T>>;
+      view.context.$implicit = record.item;
     });
 
-    // We have nulled-out all of the views that were removed or moved from previousViews. What is
-    // left is the unchanged items that we queue up to be re-inserted.
-    i = previousViews.length;
+    // Update the context variables on all items.
+    let i = this._viewContainerRef.length, count = this._data.length;
```

Comment:
I don't think we do these kinds of declarations anywhere else in the codebase.

Reply:
We don't, but this is a very performance-sensitive part, so we want to be as efficient as possible.

Reply:
I think he mean't the comma definition.